wolfcrypt/src/port/iotsafe/iotsafe.c

```diff
@@ -0,0 +1,1064 @@
+                *result = 0;
+                ret = 0;
+            } else {
+                ret = WC_HW_E; /* IoT-safe bus error */
+            }
+        }
+        else if (ret == 0) {
+            *result = 1;
+            ret = 0;
+        }
+
+        /* Terminate sign/verify session. */
+        iotsafe_cmd_start(csim_cmd, IOTSAFE_CLASS, IOTSAFE_INS_VERIFY_INIT, 1, 1);
+        iotsafe_cmd_complete(csim_cmd);
+        expect_ok(csim_cmd, (word32)strlen(csim_cmd));
+    } else {
+        /* TODO: RSA */
+        ret = -1;
+    }
+    return ret;
+}
+
+
+/*
+ * Callbacks for IoT-Safe ECC/ECDH functions
+ *
+ *
+ */
+
+#ifdef HAVE_PK_CALLBACKS
+static int wolfIoT_ecc_keygen(WOLFSSL* ssl, struct ecc_key* key,
+    unsigned int keySz, int ecc_curve, void* ctx)
+{
+    int ret;
+    (void)ctx;
+    IOTSAFE *iotsafe = &ssl->iotsafe;
+    WOLFSSL_MSG("IOTSAFE: Called wolfIoT_ecc_keygen");
+    if (iotsafe->enabled) {
+        ret = iotsafe_gen_keypair(iotsafe->privkey_id);
+        if (ret < 0)
+            return ret;
+        ret = iotsafe_get_public_key(iotsafe->privkey_id, key);
+    } else {
+        WC_RNG    rng;
+        ret = wc_InitRng(&rng);
+        if (ret != 0)
+            return ret;
+        ret = wc_ecc_init(key);
+        if (ret == 0) {
+            /* create new key */
+            ret = wc_ecc_make_key_ex(&rng, keySz, key, ecc_curve);
+        }
+        wc_FreeRng(&rng);
+    }
+    return ret;
+}
+
+
+static int wolfIoT_ecc_sign(WOLFSSL* ssl,
+       const unsigned char* in, unsigned int inSz,
+       unsigned char* out, word32* outSz,
+       const unsigned char* keyDer, unsigned int keySz,
+       void* ctx)
+{
+    int ret;
+    (void)ctx;
+    IOTSAFE *iotsafe = &ssl->iotsafe;
+    WOLFSSL_MSG("IOTSAFE: Called wolfIoT_ecc_sign\n");
+
+    if (iotsafe->enabled) {
+        ret = iotsafe_sign_hash(iotsafe->privkey_id, IOTSAFE_HASH_SHA256, IOTSAFE_SIGN_ECDSA,
+                in, inSz, out, outSz);
+        if (ret > 0)
+            return 0;
+        else
+            return ret;
+    } else {
+        WC_RNG    rng;
+        word32    idx = 0;
+        ecc_key   myKey;
+        byte*     keyBuf = (byte*)keyDer;
+        ret = wc_InitRng(&rng);
+        if (ret != 0)
+            return ret;
+
+        ret = wc_ecc_init(&myKey);
+        if (ret == 0) {
+            ret = wc_EccPrivateKeyDecode(keyBuf, &idx, &myKey, keySz);
+            if (ret == 0) {
+                ret = wc_ecc_sign_hash(in, inSz, out, outSz, &rng, &myKey);
+            }
+            wc_ecc_free(&myKey);
+        }
+        wc_FreeRng(&rng);
+    }
+    return ret;
+}
+
+static int wolfIoT_ecc_verify(WOLFSSL *ssl,
+       const unsigned char* sig, unsigned int sigSz,
+       const unsigned char* hash, unsigned int hashSz,
+       const unsigned char* keyDer, unsigned int keySz,
+       int* result, void* ctx)
+{
+    byte sig_raw[IOTSAFE_ECC_KSIZE* 2];
+    int ret;
+    ecc_key key;
```

Comment:
Might consider a WOLFSSL_SMALL_STACK option for the `ecc_key` structs in this function.

Reply:
excellent suggestion. I have added the support for WOLFSSL_SMALL_STACK in these three functions. In this one in particular also the buffer for the signature will be XMALLOC'd when WOLFSSL_SMALL_STACK is on, saving extra 64 Bytes.